### PR TITLE
Implement acceptance criterias

### PR DESCRIPTION
**Pull request by ardevelop**
@danschultz PTAL

We split PR into commits that correspond to  implementing each acceptance criteria.
Demo page: https://ardevelop.github.io/pre-flexit-training

### CODE REVIEW

**Comment by ardevelop** [comment]
Next plan:
- Optimize new state creation, we're going to clone only affected objects to reduce memory consumption.
- Add more intuitive controls for Send Forward and Send Backward actions (e.g. custom context menu)
- Smarter logic for changing z-Index to consider elements that are overlapped.
- Add ability to select and outline elements on canvas, to be able to delete them and move by keyboard.

File: deploy.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+git checkout dev
+git branch -D gh-pages
+git checkout -b gh-pages dev
+npm install react react-dom
```
Comment (danschultz) [inline deploy.sh]: assuming your `package.json` exists, you should just be able to run `npm install`

File: src/utils/model/base_model.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import {IClonable} from "../../interfaces/clonable";
+
+export type UpdatablePropeties = BaseModel | any[];
+
+export abstract class BaseModel implements IClonable<BaseModel> {
+  abstract clone(): BaseModel;
+
+  updateAll(predicate: (model: UpdatablePropeties) => UpdatablePropeties): UpdatablePropeties {
```
Comment (danschultz) [inline src/utils/model/base_model.ts]: typescript supports [polymorphic this](https://github.com/Microsoft/TypeScript/pull/4910). so you can just do `updateAll(..): this`;

File: src/utils/model/base_model.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+          cloned = this.clone();
+        }
+        (<any>cloned)[propertyName] = newValue;
+      }
+    }
+
+    return cloned || this;
+  }
+
+  update<T extends UpdatablePropeties>(model: T, func: (model: T) => T): UpdatablePropeties {
```
Comment (danschultz) [inline src/utils/model/base_model.ts]: typescript supports [polymorphic this](https://github.com/Microsoft/TypeScript/pull/4910). so you can just do `update(..): this`;

File: src/redux/canvas.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    });
+
+    return dispatcher;
+  }
+
+  add(payload: AddAction) { this.call(ADD, payload); }
+  move(payload: ElementMoveAction) { this.call(MOVE, payload); }
+  sendForward(payload: ElementAction) { this.call(SEND_FORWARD, payload); }
+  sendBackward(payload: ElementAction) { this.call(SEND_BACKWARD, payload); }
+}
```
Comment (danschultz) [inline src/redux/canvas.ts]: this is an interesting pattern that i haven't seen before in redux apps. it lgtm, but i'm curious how it scales as the app state gets more complicated.

File: src/redux/canvas.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+export default CanvasDispatcher.default;
```
Comment (danschultz) [inline src/redux/canvas.ts]: using this global throughout your app makes mocking and testing very difficult. i'd recommend instantiating instances for your dispatchers in the entry point of your app, and passing them down into your components.

**Comment by danschultz** [comment]
👍  LGTM

**Comment by ardevelop** [comment]
@danschultz PTAL
We've addressed your remarks, also added new functionality like selecting one or more (via Shift key) elements on canvas and ability to move them all.

File: src/components/app/canvas/canvas.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+          <g id="selection-elements">
+            {
+              canvas.getSelectedElements().map((e) => {
+                return (
+                  <rect key={`selection-${e.id}`} id={`selection-elements-${e.id}`}
+                      x={e.position.x - 2} y={e.position.y - 2} width={e.width + 3} height={e.height + 3} />
+                );
+              })
+            }
+          </g>
```
Comment (danschultz) [inline src/components/app/canvas/canvas.tsx]: consider putting this into it's own method and/or component.

File: src/components/app/canvas/canvas.tsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+          {
+            outterSelection.map((r => {
+              return (
+                <rect key="selection" id="selection-outter-rect"
+                      x={r.position.x - 5} y={r.position.y - 5} width={r.width + 10} height={r.height + 10} />
+              );
+            }))
+          }
+        </g>
```
Comment (danschultz) [inline src/components/app/canvas/canvas.tsx]: consider putting the selection into it's own method and/or component.

File: src/components/app/canvas/element/element.tsx [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    const {element} = this.props;
+    const {elementNode} = this.refs;
+
+    elementNode.classList.add("dragging");
+
+    let selectedElements: HTMLElement[] = [];
+    let selectedElementsOutlines: HTMLElement[];
+    let shiftKey = e.shiftKey;
+
+    new DragSession(
```
Comment (danschultz) [inline src/components/app/canvas/element/element.tsx]: consider putting the drag behavior into a new selection component or maybe the canvas component. putting the logic here violates the principle that a component only modifies itself.

File: src/app.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+export abstract class AppChildComponent<TProps> extends Component<TProps, {}> {
+  static get contextTypes() {
+    return contextType;
+  }
+
+  get app(): App {
+    return (<ContextType>this.context).app;
+  }
+}
```
Comment (danschultz) [inline src/app.ts]: 👍

File: src/dispatchers/canvas.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    });
+  }
+
+  add(payload: AddAction) { this.call(ADD, payload); }
+  move(payload: ElementMoveAction) { this.call(MOVE, payload); }
+  sendForward(payload: ElementAction) { this.call(SEND_FORWARD, payload); }
+  sendBackward(payload: ElementAction) { this.call(SEND_BACKWARD, payload); }
+  select(payload: ElementSelectAction) { this.call(SELECT, payload); }
+  selectFlush() { this.call(SELECT_FLUSH); }
+  selectMove(payload: MoveAction) { this.call(SELECT_MOVE, payload); }
```
Comment (danschultz) [inline src/dispatchers/canvas.ts]: Super minor, but I think the DX would be improved if you specified the args for each action, and the function created the action from those args.

**Comment by danschultz** [comment]
@ardevelop added some additional comments. this PR is getting a bit large. could you merge this and submit a new one with additional changes?

**Pull request merged**